File: Android/app-remote-config/AppRemoteConfig/skipstone/AppRemoteConfig/src/main/kotlin/app/remote/config/Platform.kt
package app.remote.config

import skip.lib.*
import skip.lib.Array

import skip.foundation.*

enum class Platform(override val rawValue: String, @Suppress("UNUSED_PARAMETER") unusedp: Nothing? = null): CaseIterable, RawRepresentable<String> {
    iOS("iOS"),
    iOS_iPhone("iOS.iPhone"),
    iOS_iPad("iOS.iPad"),
    iOS_tv("iOS.TV"),
    iOS_carplay("iOS.CarPlay"),
    iOS_mac("iOS.Mac"),
    macOS("macOS"),
    watchOS("watchOS"),
    visionOS("visionOS"),
    android("Android"),
    android_phone("Android.phone"),
    android_tablet("Android.tablet"),
    android_tv("Android.TV"),
    wearOS("WearOS"),
    unknown("unknown");

    internal fun applies(to: Platform): Boolean {
        val other = to
        // self can be ios
        // other can be ios.iphone
        // then match

        // self can be ios.iphone
        // other can be ios
        // then DO NOT match

        return when (this) {
            Platform.iOS -> other.rawValue.hasPrefix(Platform.iOS.rawValue)
            Platform.android -> other.rawValue.hasPrefix(Platform.android.rawValue)
            Platform.unknown -> false
            else -> this == other
        }
    }

    companion object {
        val allCases: Array<Platform>
            get() = arrayOf(iOS, iOS_iPhone, iOS_iPad, iOS_tv, iOS_carplay, iOS_mac, macOS, watchOS, visionOS, android, android_phone, android_tablet, android_tv, wearOS, unknown)
    }
}

fun Platform(rawValue: String): Platform? {
    return when (rawValue) {
        "iOS" -> Platform.iOS
        "iOS.iPhone" -> Platform.iOS_iPhone
        "iOS.iPad" -> Platform.iOS_iPad
        "iOS.TV" -> Platform.iOS_tv
        "iOS.CarPlay" -> Platform.iOS_carplay
        "iOS.Mac" -> Platform.iOS_mac
        "macOS" -> Platform.macOS
        "watchOS" -> Platform.watchOS
        "visionOS" -> Platform.visionOS
        "Android" -> Platform.android
        "Android.phone" -> Platform.android_phone
        "Android.tablet" -> Platform.android_tablet
        "Android.TV" -> Platform.android_tv
        "WearOS" -> Platform.wearOS
        "unknown" -> Platform.unknown
        else -> null
    }
}

File: Sources/AppRemoteConfig/Platform.swift
import Foundation

/// Lists supported platforms
public enum Platform: String, CaseIterable {
    /// iOS in any of its variants
    case iOS = "iOS"
    
    /// iOS on an iPhone
    case iOS_iPhone = "iOS.iPhone"
    
    /// iOS on an iPad
    case iOS_iPad = "iOS.iPad"
    
    /// iOS on an Apple TV
    case iOS_tv = "iOS.TV"
    
    /// iOS on CarPlay
    case iOS_carplay = "iOS.CarPlay"
    
    /// iOS on an a Mac using Catalyst
    case iOS_mac = "iOS.Mac"
    
    /// Linux
    case linux = "Linux"
    
    /// macOS
    case macOS = "macOS"
    
    /// watchOS
    case watchOS = "watchOS"
    
    /// visionOS
    case visionOS = "visionOS"
    
    /// Android in any of its variants
    case android = "Android"
    
    /// Android on a phone
    case android_phone = "Android.phone"
    
    /// Android on a table
    case android_tablet = "Android.tablet"
    
    /// Android on a tv
    case android_tv = "Android.TV"
    
    /// WearOS
    case wearOS = "WearOS"
    
    /// Windows
    case windows = "Windows"
    
    /// Unknown
    case unknown
    
    func applies(to other: Platform) -> Bool {
        // self can be ios
        // other can be ios.iphone
        // then match
        
        // self can be ios.iphone
        // other can be ios
        // then DO NOT match
        
        switch self {
        case .iOS:
            other.rawValue.hasPrefix(Platform.iOS.rawValue)
        case .android:
            other.rawValue.hasPrefix(Platform.android.rawValue)
        case .unknown:
            false
        default:
            self == other
        }
    }
}
